### PR TITLE
Support bundling HTML files and their js, css, and assets in `Bun.build` and `bun build`

### PR DESCRIPTION
**Pull request by Jarred-Sumner**
### What does this PR do?

Fixes https://github.com/oven-sh/bun/issues/4688

This lets you bundle HTML files in Bun. It uses `lol-html` to parse and enqueue assets.

Currently, elements matching the following selectors are considered for either bundling (if js/css) or copied when non-external:
- `script[src]`
- `link[rel='stylesheet'][href]`
- `link[as='style'][href]`
- `link[as='font'][href], link[type^='font/'][href]`
- `link[as='image'][href]`
- `link[as='video'][href], link[as='audio'][href]`
- `link[as='worker'][href]`
- `link[rel='manifest'][href]`
- `link[rel='icon'][href], link[rel='apple-touch-icon'][href]`
- `link:not([rel~='stylesheet']):not([rel~='modulepreload']):not([rel~='manifest']):not([rel~='icon']):not([rel~='apple-touch-icon'])[href]`
- `img[src]`
- `img[srcset]`
- `video[src]`
- `video[poster]`
- `audio[src]`
- `source[src]`
- `source[srcset]`
- `iframe[src]`

This list will probably change before this merges. It should probably be configurable.

### How did you verify your code works?

There are tests.

### CODE REVIEW

**Comment by robobun** [comment]
<div><sup>Updated 4:28 AM PT - Dec 22nd, 2024</sup></div>

@Jarred-Sumner, your commit 2abfa554086e74170e45f749637826e5f1d53660 is building: [`#8416`](https://buildkite.com/bun/bun/builds/8416)

File: src/bake/DevServer.zig [hunk 1]
```diff
@@ -9,7 +9,7 @@
 pub const DevServer = @This();
 pub const debug = bun.Output.Scoped(.Bake, false);
 pub const igLog = bun.Output.scoped(.IncrementalGraph, false);
-
+const Chunk = @import("../bundler//bundle_v2.zig").Chunk;
```
Comment (paperclover) [inline src/bake/DevServer.zig]: avoid file imports if possible

const Chunk = bun.bundle_v2.Chunk

(move this at the end of the file)

File: src/bundler/bundle_v2.zig [hunk 2]
```diff
@@ -6036,6 +6113,12 @@ pub const LinkerContext = struct {
                 }
             }
 
+            if (experimental_html) {
+                for (html_chunks.values()) |*chunk| {
```
Comment (paperclover) [inline src/bundler/bundle_v2.zig]: appendSliceAssumeCapacity

File: src/bundler/bundle_v2.zig [hunk 3]
```diff
@@ -9389,6 +9484,163 @@ pub const LinkerContext = struct {
...
+        //    the JavaScript for the entry point which uses the "src" attribute
+        //    to point to the JavaScript chunk's unique key.
+        // 4. If a CSS chunk exists, add a <link rel="stylesheet" href="..." crossorigin> tag that contains
+        //    the CSS for the entry point which uses the "href" attribute to point to the
+        //    CSS chunk's unique key.
+        // 5. For each imported module or chunk within the JavaScript code, add
+        //    a <link rel="modulepreload" href="..." crossorigin> tag that
+        //    points to the module or chunk's unique key so that we tell the
+        //    browser to preload the user's code.
+        // We might also want to force <!DOCTYPE html> at the top of the file, but I'm not sure about that yet.
```
Comment (paperclover) [inline src/bundler/bundle_v2.zig:9590]: its fine to take no action re:doctype

the one person who wants to bundle quirks mode html can do so

File: src/bundler/bundle_v2.zig [hunk 4]
```diff
@@ -9389,6 +9484,163 @@ pub const LinkerContext = struct {
...
+            pub fn onHTMLParseError(this: *@This(), message: []const u8) void {
+                this.log.addError(
+                    &this.linker.parse_graph.input_files.items(.source)[this.source_index],
+                    Logger.Loc.Empty,
+                    message,
+                ) catch bun.outOfMemory();
+            }
+
+            pub fn onTag(this: *@This(), element: *lol.Element, path: []const u8, url_attribute: []const u8, kind: ImportKind) void {
+                _ = kind; // autofix
```
Comment (paperclover) [inline src/bundler/bundle_v2.zig]: no autofix comments pls

remove the comment or make the parameter name _

File: src/html_scanner.zig [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+                },
+                // Icons
+                .{
+                    .selector = "link[rel='icon'][href], link[rel='apple-touch-icon'][href]",
+                    .url_attribute = "href",
+                    .kind = .url,
+                },
+                // Catch-all for other links with href
+                .{
+                    .selector = "link:not([rel~='stylesheet']):not([rel~='modulepreload']):not([rel~='manifest']):not([rel~='icon']):not([rel~='apple-touch-icon'])[href]",
```
Comment (paperclover) [inline src/html_scanner.zig]: not 100% sure if this is a good idea.

File: src/html_scanner.zig [hunk 6]
```diff
@@ -0,0 +1,297 @@
+const std = @import("std");
```
Comment (paperclover) [inline src/html_scanner.zig]: whenever a file with a single `struct` is added, make that struct top-level and make the file name the name of that struct (`html_scanner.zig` to `HTMLScanner.zig`). when i do these, i then make one of the first line the text `const StructName = @This()` directly followed by the fields.

this pattern pairs really well with imports at the end of the file, so the start reads well

File: test/bundler/bundler_html.test.ts [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+      expect(page2CssPath).toBeDefined();
+
+      const page1Css = api.readFile("out/" + page1CssPath!);
+      const page2Css = api.readFile("out/" + page2CssPath!);
+      expect(page1Css).toContain("box-sizing:border-box");
+      expect(page2Css).toContain("box-sizing:border-box");
+      expect(page1Css).toContain(".shared");
+      expect(page2Css).toContain(".shared");
+    },
+  });
```
Comment (paperclover) [inline test/bundler/bundler_html.test.ts:469]: can a test that has a js entrypoint importing an html file be added

additionally a js file importing an html file importing the first js file

i assume these already work but would be good to have it tested

**Review by paperclover — CHANGES_REQUESTED** [review]
overall legendary pr. very nice to have this.

random nitpicks and two that should be considered before merging. the main logic looks fine

**Comment by ohroy** [comment]
awesome !!! but would you fix https://github.com/oven-sh/bun/issues/5241 ? this block we switch to bun build

File: src/bake/DevServer.zig (hunk 1 above)
Comment (Jarred-Sumner) [reply src/bake/DevServer.zig]: sounds like a linter rule!

File: src/html_scanner.zig (hunk 5 above)
Comment (Jarred-Sumner) [reply src/html_scanner.zig]: we rewrite all links this way

File: src/html_scanner.zig (hunk 6 above)
Comment (Jarred-Sumner) [reply src/html_scanner.zig]: There are about 5 places we do this currently

I'm not opposed to it

but these patterns are really only good when they're consistent

**Comment by marcospgp** [comment]
awesome addition!

**Pull request merged**